### PR TITLE
Fix clip_grad_norm_ failure with mixed DeviceMesh DTensor norms

### LINKED ISSUE #180346: [DTensor] clip_grad_norm_ fails when per-parameter norms come from different DeviceMesh objects
### 🚀 The feature, motivation and pitch

`torch.nn.utils.clip_grad._get_total_norm()` groups tensors by (device, dtype) only, computes per-tensor norms, and then stacks those norms together. For DTensor inputs, the norms themselves remain DTensors and retain mesh metadata, even if such tensors are scalars ultimately. If the norms come from different DeviceMesh objects, the final `torch.stack(...)` fails even though the tensors share the same device and dtype.

Expected: `clip_grad_norm_` handles mixed meshes correctly. I considered extracting DTensor scalar norms via `item()`, but that only seems safe when each norm is known to be a replicated scalar. For general DTensor layouts, `item()` reads the local value and can be numerically wrong, so it does not look like a sound generic fix.

Actual: the generic `torch.stack` path fails because the norm DTensors carry different mesh metadata. This follows directly from the current `_get_total_norm()` implementation.

### Alternatives

_No response_

### Additional context

This seems distinct from the already reported performance issue in the same code path (https://github.com/pytorch/pytorch/issues/169445): this report is about a runtime failure with mixed meshes, not about stack complexity.

Related https://github.com/pytorch/pytorch/issues/121020.

Reproduction
```python
import torch
import torch.distributed as dist
from torch.distributed._tensor import DTensor, Replicate
from torch.distributed.device_mesh import DeviceMesh
from torch.testing._internal.common_distributed import spawn_threads_and_init_comms

WORLD_SIZE = 2

@spawn_threads_and_init_comms(world_size=WORLD_SIZE)
def repro_mixed_mesh_clip_grad_norm_failure(world_size: int) -> None:
    rank = dist.get_rank()

    mesh_a = DeviceMesh("cpu", torch.arange(world_size))
    mesh_b = DeviceMesh("cpu", torch.arange(world_size - 1, -1, -1))

    placements = [Replicate()]

    local_a = torch.tensor([3.0 + rank], dtype=torch.float32)
    local_b = torch.tensor([4.0 + rank], dtype=torch.float32)

    dt_a = DTensor.from_local(local_a, mesh_a, placements)
    dt_b = DTensor.from_local(local_b, mesh_b, placements)

    norm_type = 2.0
    norms = [torch.linalg.vector_norm(g, norm_type) for g in [dt_a, dt_b]]

    first_device = dt_a.device
    total_norm = torch.linalg.vector_norm(
        torch.stack([norm.to(first_device) for norm in norms]),
        norm_type,
    )
    print(rank, total_norm)

if __name__ == "__main__":
    repro_mixed_mesh_clip_grad_norm_failure(WORLD_SIZE)
```

cc @awgu @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @pragupta @msaroufim @dcci @aditvenk @xmfan @weifengpy

### PR DESCRIPTION
**Pull request by Charly21r**
## Summary

Fixes #180346 

This PR fixes two related failures in `clip_grad_norm_` when gradients are DTensors from different `DeviceMesh` objects:

**1. `torch.stack` on mixed-mesh norms in `_get_total_norm()`**

Per-parameter norms are stacked via `torch.stack()`. When the norms are DTensors from different meshes, `torch.stack` cannot reconcile the mesh metadata. Fix: materialise DTensor norms to plain tensors via `full_tensor()` before stacking, but only when the norms actually come from different meshes. The common single-mesh path (FSDP, TP) is untouched.

**2. `_foreach_mul_.Tensor` on mixed-mesh grads in `_clip_grads_with_norm_()`**

The clipping coefficient is multiplied across all gradients via `_foreach_mul_`. The `.Tensor` overload mesh-tags the scalar, which fails when grads span different meshes. Fix: when the grads share one mesh, keep the existing tensor-coefficient fast path (avoiding a CPU↔device sync). When they don't, coerce the coefficient to a Python float via `.item()` so dispatch goes to the `.Scalar` overload, which carries no mesh metadata. The foreach fast path is preserved in both cases.

Both call sites use a shared `_has_mixed_dtensor_meshes()` helper that uses value-equality (`!=`) on the `DeviceMesh` objects, so different `DeviceMesh` instances representing the same mesh don't trigger unnecessary materialisation.

### Benchmark (256 gradients, single rank, CPU)

| Case | Before | After |
|---|---|---|
| Plain tensors | 75 µs | 85 µs |
| Single mesh DTensors | 250 µs | 269 µs |
| Mixed mesh DTensors | **CRASH** | 261 µs |

## Test plan

- [x] New test `test_clip_grad_norm_mixed_mesh` in `test/distributed/tensor/test_dtensor.py` covering both `Replicate()` and `Shard(0)` placements across two different `DeviceMesh` objects
- [x] Verified plain tensor path is unaffected
- [x] Verified single-mesh DTensor path is unaffected
- [x] Benchmarked against naive fix (always `full_tensor()`)
- [x] Verified `clip_grad_value_` is not affected by the same class of bug, it already passes a Python float, dispatching to the `.Scalar` overload.

### CODE REVIEW

**Review by stmcgovern — COMMENTED** [review]
This review was generated by claude. It is just a community review and its findings are only informational. 

PR Review: #180501                                                                                                                                     
                                                                                                                                                         
  Summary                                                                                                                                                
   
  The _get_total_norm fix is correct and well-targeted. The _clip_grads_with_norm_ changes work but can be simplified, and as-written introduce a subtle 
  logic issue in the foreach=True error path.
                                                                                                                                                         
  Code Quality                                                                                                                                           
   
  _get_total_norm fix is correct — torch.stack requires mesh agreement. Stripping DTensor metadata via full_tensor() at the point where a distributed    
  gradient becomes a scalar norm is the right boundary.
                                                                                                                                                         
  _clip_grads_with_norm_ can be simplified — The _foreach_mul_ failure happens because the .Tensor variant mesh-tags the scalar clip_coef_clamped, but   
  the .Scalar variant (Python float) has no mesh to conflict. You still need _has_mixed_dtensor_meshes detection, but instead of falling back to a
  per-element Python loop, convert the coefficient to a float:                                                                                           
                  
  if use_foreach and not mixed_meshes:
      torch._foreach_mul_(device_grads, clip_coef_clamped.to(device))
  elif use_foreach and mixed_meshes:                                                                                                                     
      torch._foreach_mul_(device_grads, clip_coef_clamped.to(device).item())
                                                                                                                                                         
  This preserves the foreach fast path in both cases. The .item() call adds a CPU-GPU sync, but only in the mixed-mesh path — the common single-mesh path
   is untouched, and mixed-mesh is already doing full_tensor() collectives in _get_total_norm. This also sidesteps the elif issue below.
                                                                                                                                                         
  elif logic issue — The change from elif foreach: to elif foreach and not use_foreach: silently swallows explicit foreach=True when mixed meshes are    
  detected. Trace: foreach=True, device supports foreach → use_foreach=True → first branch skipped (mixed) → elif foreach and not use_foreach → True and 
  False → skipped → silent scalar fallback with no warning.                                                                                              
                  
  Testing

  - Add a Shard placement case — the test uses Replicate() where full_tensor() is nearly a no-op. Shard(0) would exercise the collective path.           
  - clip_grad_value_ uses _foreach_clamp_{min,max}_ which has the same .Tensor variant issue with mixed meshes — worth a follow-up.

**Comment by Charly21r** [comment]
Fixed `_clip_grads_with_norm_`, passing `.item()` to `_foreach_mul_` keeps the foreach fast path for mixed meshes without the `.Tensor` mesh-tagging conflict. Also expanded the test to cover `Shard(0)`.

Verified locally with different meshes, `clip_grad_value_` is not affected. It passes `clip_value` as a Python float (after `clip_value = float(clip_value)`), so `_foreach_clamp_min_` / `_foreach_clamp_max_` dispatch to the `.Scalar` overload, which doesn't tag the bound with mesh metadata. I also confirmed `_foreach_clamp_min_.Tensor` succeeds across mixed meshes anyway, it doesn't have the same cross-mesh validation that `_foreach_mul_.Tensor` does. So no follow-up needed.

**Comment by Charly21r** [comment]
Hi! Friendly follow-up on this PR.

I addressed the review feedback (including the mixed-mesh foreach path and Shard(0) coverage), and all tests are still passing locally.

Would appreciate another look when someone has bandwidth. Thanks!

File: torch/nn/utils/clip_grad.py [hunk 1]
```diff
@@ -167,10 +195,17 @@ def _clip_grads_with_norm_(
...
+        use_foreach = (
+            foreach is None and _has_foreach_support(device_grads, device)
+        ) or (foreach and _device_has_foreach_support(device))
+        if use_foreach:
+            coef = clip_coef_clamped.to(device)
+            # _foreach_mul_.Tensor tags the scalar with mesh metadata and
+            # fails when grads span different meshes.  Passing a Python float
+            # via .item() avoids the conflict while keeping the foreach path.
+            if _has_mixed_dtensor_meshes(device_grads):
+                coef = coef.item()
```
Comment (wconstab) [inline torch/nn/utils/clip_grad.py:207]: i'm not sure about adding a .item call here. It forces a sync and breaks graph tracing, among other things.

I also worry it's a slippery slope: we try to be clear about the requirements/limitations for dtensor operations- same mesh, except for a special case for foreach operators which are themselves loops over separate tensors that can have distinct meshes.

The alternative would be to deal with this in the trainer layer, where there is knowledge of the multiple-mesh situation.

Comment (Charly21r) [reply torch/nn/utils/clip_grad.py:207]: Good point on `.item()`, I'll drop it. 

For the mixed-mesh case I can just skip the foreach fast path and fall back to the existing per-grad loop (`for g in device_grads: g.mul_(coef)`). Each call only touches one mesh, so there's no cross-mesh dispatch, and no sync or tracing concern since `.item()` is gone. The change is a single extra precondition on `use_foreach`, no new code paths.

```python
use_foreach = (
    (foreach is None and _has_foreach_support(device_grads, device))
    or (foreach and _device_has_foreach_support(device))
) and not _has_mixed_dtensor_meshes(device_grads)
```

Happy to also revisit the `_get_total_norm` change if you'd prefer that surface a clear error instead, but `full_tensor()` there acts on per-grad norm scalars, which felt like a reasonable use of a public DTensor API. Let me know which way you'd like me to go.

Comment (wconstab) [reply torch/nn/utils/clip_grad.py:207]: One more question- would you prefer to degenerate to no foreach here, or would you prefer to keep the foreach for performance but break up the list of grads by mesh and do a few separate foreaches? It seems like the outcome of totally dropping it might be worse than having an error that forces a change in the call site, but I'm not sure.

Comment (Charly21r) [reply torch/nn/utils/clip_grad.py:207]: Good question, I'd lean towards keeping foreach but grouping by mesh, so you'd get one `_foreach_mul_` per distinct mesh. That avoids the silent-perf-trap concern (no degeneration to Python loop) while still working for mixed-mesh callers:

```python
if _has_mixed_dtensor_meshes(device_grads):
    for mesh_grads in _group_by_mesh(device_grads):
        torch._foreach_mul_(mesh_grads, coef)
else:
    torch._foreach_mul_(device_grads, coef)
```

`_group_by_mesh` would partition by `DTensor.device_mesh` (value-equal), with plain tensors as their own bucket. Perf for mixed-mesh becomes N small foreach calls instead of 1, which should be close to original.

If the explicit-error path is preferred, I can go that way too.
